### PR TITLE
Swipe to delete

### PR DESCRIPTION
**Pull request by ldematte**
I took a shot at issue #3354 (swipe to delete).

For now, it implements the swiping gestures and events (inside ConversationListItem, following the same pattern used for click and long-press). It replaces the "old" click and long-press listeners, as swiping has complex interactions with touch events handling (and handling all of them in one place make it more consistent and easier). Gestures are modelled after the GestureDetector class; specific cases like multi-finger touches are handled too.

It also plays well with the parent RecycleView scrolling; I tried to make it work nicely and avoid to subclass it and it seems to have worked.
It reproduces as closely as possible gmail behaviour.

What is missing, and IMO separate, broader issues, is what to do after swiping: 
- delete, "archive", or let the user choose (with a "do nothiing" option too)? (Through preferences, maybe?) (right now both swipe-to-left and swipe-to-right do a "delete")
- Undo: right now it is just a delay for delete: an item is scheduled for delete in 5 seconds, if users press "Undo", deletion is cancelled. An interesting alternative is to do not actually delete the message(s): alter MessagingDatabase tables to mark an item as deleted (add something like a "delete time"); then, periodically sweep tables.
- This second option may be the starting point for an "archive" or recycle bin functionality (which I believe would be an entirely new (interesting) feature/issue).

I tried to stay as close to your coding style as possible.
Let me know what you think!

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -32,6 +32,9 @@ repositories {
     maven { // cwac-camera
         url 'https://repo.commonsware.com.s3.amazonaws.com'
     }
+    maven { //material-dialogs
```
Comment (agrajaghh) [inline build.gradle]: this shouldn't be in this PR I guess

File: build.gradle [hunk 2]
```diff
@@ -209,6 +211,7 @@ android {
                           'proguard-shortcutbadger.pro',
                           'proguard-retrofit.pro',
                           'proguard.cfg'
+            applicationIdSuffix ".debug"
```
Comment (agrajaghh) [inline build.gradle]: same here

File: build.gradle (hunk 1 above)
Comment (ldematte) [inline build.gradle]: I only modified the build.gradle to make it compile with com.afollestad:material-dialogs (there was a dep problem, already pointed out in a separate issue).
I probably should exclude the whole file from the PR, thanks

**Comment by TheNain38** [comment]
Would there be a new setting to disable that behavior and get the old one ?

**Comment by agrajaghh** [comment]
> the answer is never more options

and what old behaviour? Nothing happens when you swipe...

**Comment by TheNain38** [comment]
Yes, but if you mistakely swipe when you did not meant to, then it could be annoying

**Comment by ldematte** [comment]
(I take gmail as an example here as a very used application with swiping features, but the behaviour I have observed is quite consistent).

There is an undo option, like suggested in the issue and like in gmail. So if you did it by mistake, and you realize it right away, you can undo it.
I see three options here:
1. configuration option: choose action for swipe left/right (nothing, delete, archive) (gmail does that)
Note that the third option requires extensive work to Signal: there is no archive atm.
2. swiping (deletion, in general!) will move messages to a recycle bin, instead of directly deleting them (also like in gmail). Very easy.. if you add a recycle bin to Signal!
3. Add a confirmation message.

Personally, I like 1. and 2.
I think they are both separate issues, but very complementary to this PR, and will be great to have them. 
Personally, I think that this PR plus 1. (with safe defaults: right to archive, left to none) and 2. will make a great UX.
If the WhisperSystem dev team likes the current PR, I would love to open issues for 1. and 2. and add them to the app. If they reject it... what's the purpose? (well, n.2 has a great value on its own, TBH)
I hate n.3, but this is OSS, so if someone else wants to add it, free to work on it ;)

PS: I understand "the answer is never more options" , but my feeling is that this advice is more related to the encryption/security aspects of Signal. But if you really hate options, than a recycle bin/archive is the answer.

**Comment by TheNain38** [comment]
For example: gmail is annoying me with this because, sometimes I meant to go up or down and instead it was deleting the email -_-

**Comment by agrajaghh** [comment]
for reference: in #1141 there was already some work done on archive..

**Comment by ldematte** [comment]
@TheNain38 gmail has options to make a swipe go to archive  (default) or disable it completely. That's what n.1 was about.

@agrajaghh: I think #1141 plus this would be great!

**Comment by TheNain38** [comment]
@ldematte Ok, thanks

**Comment by ldematte** [comment]
@TheNain38 glad to help!

**Comment by moxie0** [comment]
Thanks we have swipe to archive in now, and are pausing to evaluate whether we'd like to include swipe to delete.  If we do, we'll probably use the recyclerview stuff rather than the gesture stuff.  Thanks!

**Comment by Stathread** [comment]
Please re-open this. I am about to un-install Signal for this reason alone. I don't want to archive my messages, I want to delete them. I am a user of Signal and like the Google swipe to delete.

Obviously you wouldn't make this default. You put a setting in to allow people to switch from "Swipe to Archive" to "Swipe to Delete" that way everyone complaining about "accidentally" deleting messages can stop complaining.

**Comment by dcherian** [comment]
Swipe left to delete. Swipe right to archive?

**Pull request closed without merging**